pytools/graph.py

```diff
@@ -566,24 +567,39 @@ def undirected_graph_from_edges(
 
 def get_reachable_nodes(
         undirected_graph: GraphT[NodeT],
-        source_node: NodeT) -> frozenset[NodeT]:
+        source_node: NodeT,
+        exclude_nodes: Optional[set[NodeT]] = None) -> frozenset[NodeT]:
     """
     Returns a :class:`frozenset` of all nodes in *undirected_graph* that are
-    reachable from *source_node*.
+    reachable from *source_node*. Reachability of nodes from *source_node* can
+    be restricted by supplying a set of nodes to be excluded from consideration.
+    Effectively, this amounts to traversing partitions of *undirected_graph*
+    with partition points defined by *exclude_nodes*.
```

Comment:
Rephrase this in terms of nodes and edges. I find the "partition" terminology to be unclear.